src/lib.rs

```diff
@@ -85,20 +85,24 @@ use metadata::Metadata;
 
 use errors::*;
 
+macro_rules! box_result {
+    ($exp:expr) => (Ok(Box::new($exp?)))
+}
+
 /// `fetch_metadata` is the generic, top-level function that is used by the main
 /// function to fetch metadata. The configured provider is passed in and this
 /// function dispatches the call to the correct provider-specific fetch function
-pub fn fetch_metadata(provider: &str) -> Result<Metadata> {
+pub fn fetch_metadata(provider: &str) -> Result<Box<providers::MetadataProvider>> {
```

Comment:
Why is the Box needed?

Reply:
Since we're returning multiple different concrete types that all implement `MetadataProvider`, `impl Trait` can't be used and we need a `Box` to create an actual trait object.

Reply:
Unfortunately not. From [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/1522-conservative-impl-trait.md#semantics):
>If a function returns impl Trait, its body can return values of any type that implements Trait, but all return values need to be of the same type.

Reply:
@csssuf for my own reference: pushing the `-> impl Trait` one more level down too (i.e. returned by `new()`), would that also not work?

Reply:
[Nope](https://play.rust-lang.org/?gist=6e939251c0048645127f65272cd7f875&version=stable&mode=debug). The compiler needs to be able to figure out static dispatch to use `impl Trait`, so there's no way to trick it into doing the thing we want here.

Reply:
[Nope](https://play.rust-lang.org/?gist=6e939251c0048645127f65272cd7f875&version=stable&mode=debug). The compiler needs to be able to figure out static dispatch to use `impl Trait`, so there's no way to trick it into doing the thing we want here.

Reply:
https://www.reddit.com/r/rust/comments/8sqiin/newbie_question_why_is_the_impl_and_dyn_in_the/ is also useful on this topic (there's various others too)

With rustc beta today for example, one can write:

```
diff --git a/src/lib.rs b/src/lib.rs
index 0f80d1c..2ec3af3 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -90,7 +90,7 @@ macro_rules! box_result {
 /// `fetch_metadata` is the generic, top-level function that is used by the main
 /// function to fetch metadata. The configured provider is passed in and this
 /// function dispatches the call to the correct provider-specific fetch function
-pub fn fetch_metadata(provider: &str) -> Result<Box<providers::MetadataProvider>> {
+pub fn fetch_metadata(provider: &str) -> Result<Box<dyn providers::MetadataProvider>> {
     match provider {
         "azure" => box_result!(azure::Azure::new()),
         "cloudstack-metadata" => box_result!(cloudstack::network::CloudstackNetwork::new()),
```

Then it's explicitly clear the goal is dynamic dispatch.

BTW I love how much discussion there's been over a single heap allocation :smile:

Reply:
without having looked into it at all, I wonder if it would be possible to simultaneously use `dyn` and `impl Trait` to do what we want without a heap allocation? mostly curious, the current version is totally fine by me too.

Reply:
No, they're mutually exclusive. We _could_ avoid the `Box` by using an enum as the function's return type, and essentially provide our own dynamic dispatch by implementing the trait for the enum, but IMO that's even less clean than simply returning a `Box`ed trait object.